### PR TITLE
Updated token internals to ERC721 standard

### PR DESCRIPTION
**Pull request by leoloco**

### CODE REVIEW

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 1]
```diff
@@ -1,282 +1,236 @@
...
-  Storage.set(symbolKey, symbol);
-  Storage.set(totalSupplyKey, u64ToBytes(totalSupply));
-  Storage.set(baseURIKey, baseURI);
-  Storage.set(ownerKey, Context.caller().toString());
-  Storage.set(counterKey, u64ToBytes(initCounter));
-}
+export const BALANCE_KEY_PREFIX = 'BALANCE';
+export const OWNER_KEY_PREFIX = 'OWNER';
+export const ALLOWANCE_KEY_PREFIX = 'ALLOWANCE';
+export const OPERATOR_ALLOWANCE_KEY_PREFIX = 'OPERATOR_ALLOWANCE';
```
Comment (peterjah) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:20]: personnal point of view but if we want save storage we should use shorter keys. one or 2 bytes should be enough. even more when its a prefix that will be writen for each users

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 2]
```diff
@@ -1,282 +1,236 @@
...
-  const left = nft1_ownerOf(u64ToBytes(tokenId));
-  return (
-    memory.compare(
-      changetype<usize>(left),
-      changetype<usize>(stringToBytes(address)),
-      left.length,
-    ) == 0
-  );
+function ownerKey(tokenId: u256): StaticArray<u8> {
+  return stringToBytes(OWNER_KEY_PREFIX + tokenId.toString());
```
[the remark this answers is not in the document]
Comment (peterjah) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:47]: maybe we should see what is doing this function. and maybe we can code it differently.

[the remark this answers is not in the document]
Comment (peterjah) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:47]: In fact we had already the same issue with the previous implementation, and thats why we choose to put the total suppply (and thus tokenIds) on native u64 instead of u256

[the remark this answers is not in the document]
Comment (leoloco) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:47]: Yes indeed. But, I think we should maintain compatibility with EVM types.
What do you think about we set owner prefix as bytes and instead of doing tokenId.toString we use u256toBytes(tokenId) and we concat prefix and bytes representation of tokenId ?

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 3]
```diff
@@ -1,282 +1,231 @@
...
 
-export function _isApproved(address: string, tokenId: u64): bool {
-  if (address.length === 0) return false;
-  const approvedAddress = _getApproved(tokenId);
-  return approvedAddress === address;
+/**
+ * Returns the symbol of the contract.
+ */
+export function _symbol(): string {
+  if (Storage.has(SYMBOL_KEY)) {
```
Comment (peterjah) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:110]: useless if considering this is set in contructor. isnt it ?

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 4]
```diff
@@ -1,282 +1,231 @@
...
-export function _getApproved(tokenId: u64): string {
-  const key = approvedTokenKey + tokenId.toString();
-  if (!Storage.has(key)) return '';
-
-  return Storage.get(key);
+/**
+ * Returns the name of the contract.
+ */
+export function _name(): string {
+  if (Storage.has(NAME_KEY)) {
```
Comment (peterjah) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:100]: useless if considering this is set in contructor. isnt it ?

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 5]
```diff
@@ -1,282 +1,231 @@
...
+ *
+ * @param operator - address of the operator
+ * @param tokenId - tokenId of the token
+ * @returns
+ */
+export function _isApproved(operator: string, tokenId: u256): bool {
+  const allowKey = allowanceKey(tokenId);
+  if (Storage.has(allowKey)) {
+    return bytesToString(Storage.get(allowKey)) == operator;
+  }
```
Comment (peterjah) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:152]: if i understand well , only one allowed operator is possible ?
maybe it would be more efficient to store the operator address in the key, then this function would become
```
const allowKey = allowanceKey(tokenId, operator);
return Storage.has(allowKey);

**Review by peterjah — APPROVED** [review]

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts (hunk 3 above)
Comment (leoloco) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:110]: It is indeed set in the constructor but I have no guarantee that final contract will call the super constructor, that is why I added this verification. But if you think it's irrelevant and that developers should make sure to call the super constructor then I can remove it.

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts (hunk 5 above)
Comment (leoloco) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:152]: Good point. But this imply that we should delete allowanceKey when revoking allowance

Comment (leoloco) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:152]: After all, I'm not sure it's a good idea anymore. See, if the operator address is in the key, then when calling 'approve' method, you'll have to check in the storage if an approval is not already set on this id. We would have to filter the storage based on the prefix+tokenId. The same would also apply for the _getApproved method.
I think we should keep it simple. Your suggestion implies heavy modification and complex storage access for little gas efficiency. 

But I'll keep the delete allowanceKey when approval is called with zero address as parameter

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 6]
```diff
@@ -1,282 +1,239 @@
...
-  Storage.set(symbolKey, symbol);
-  Storage.set(totalSupplyKey, u64ToBytes(totalSupply));
-  Storage.set(baseURIKey, baseURI);
-  Storage.set(ownerKey, Context.caller().toString());
-  Storage.set(counterKey, u64ToBytes(initCounter));
-}
+export const BALANCE_KEY_PREFIX: StaticArray<u8> = [0x03];
+export const OWNER_KEY_PREFIX: StaticArray<u8> = [0x04];
+export const ALLOWANCE_KEY_PREFIX: StaticArray<u8> = [0x05];
+export const OPERATOR_ALLOWANCE_KEY_PREFIX: StaticArray<u8> = [0x06];
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:19]: Personal preference: I would use an enum instead. SOmething like:
```suggestion
export enum Key {
  NAME = [0x01],
  SYMBOL = [0x02],
  BALANCE_PREFIX = [0x03],
  OWNER_PREFIX = [0x04],
  ALLOWANCE_PREFIX = [0x05],
  OPERATOR_ALLOWANCE_PREFIX = [0x06]
}
```

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 7]
```diff
@@ -1,282 +1,239 @@
...
+  return u256.Zero;
 }
 
 /**
- * Removes the approval of the token
- * @param tokenId - the tokenID
+ * Returns the owner of a given tokenId.
+ *
+ * @param tokenId - The identifier for an NFT
+ * @returns the address of the owner of the NFT
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:86]: or an empty string if tokenId is not found.

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 8]
```diff
@@ -1,282 +1,239 @@
...
+  return '';
 }
 
-export function _getApproved(tokenId: u64): string {
-  const key = approvedTokenKey + tokenId.toString();
-  if (!Storage.has(key)) return '';
-
-  return Storage.get(key);
+/**
+ * Returns the name of the contract.
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:97]: or an empty string if the contract was never initialized

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 9]
```diff
@@ -1,282 +1,239 @@
...
+  }
+  return '';
 }
 
-export function _isApproved(address: string, tokenId: u64): bool {
-  if (address.length === 0) return false;
-  const approvedAddress = _getApproved(tokenId);
-  return approvedAddress === address;
+/**
+ * Returns the symbol of the contract.
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:107]: or an empty string if the contract was never initialized

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 10]
```diff
@@ -1,282 +1,239 @@
...
-
-  Storage.set(key, approved.toString());
+/**
+ * Change or reaffirm the approved address for an NFT
+ *
+ * @param tokenId - The NFT to approve
+ * @param approved - The new approved NFT controller
+ */
+export function _approve(approved: string, tokenId: u256): void {
+  assert(_isAuthorized(Context.caller().toString(), tokenId), 'Unauthorized');
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:123]: I would export 'unauthorized' as a log token to avoid copy/paste error on the application side.

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 11]
```diff
@@ -1,282 +1,239 @@
...
+  const key = allowanceKey(tokenId);
+  if (approved == '') {
+    Storage.del(key);
+  } else {
+    Storage.set(key, stringToBytes(approved));
+  }
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:129]: Can you shortly explain this behavior in the function documentation ?

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 12]
```diff
@@ -1,282 +1,239 @@
...
+ * @param tokenId - The NFT to approve
+ * @param approved - The new approved NFT controller
+ */
+export function _approve(approved: string, tokenId: u256): void {
+  assert(_isAuthorized(Context.caller().toString(), tokenId), 'Unauthorized');
+  const key = allowanceKey(tokenId);
+  if (approved == '') {
+    Storage.del(key);
+  } else {
+    Storage.set(key, stringToBytes(approved));
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:128]: You're not checking approved format. Is it on purpose ?

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 13]
```diff
@@ -1,282 +1,239 @@
...
-
-  if (!Storage.has(key)) return false;
-
-  const approved = Storage.get(key);
-
-  return approved == 'true';
+/**
+ * Get the approved address for a single NFT
+ * @param tokenId - Id of the NFT
+ * @returns Address of the approved owner of the NFT
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:135]: or an empty string.

Why not returning an address object instead ?

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 14]
```diff
@@ -1,282 +1,239 @@
...
-// ==================================================== //
-
-export function assertIsMinted(tokenId: u64): void {
-  assert(_onlyMinted(tokenId), `Token ${tokenId.toString()} is not minted`);
+/**
+ * Returns the allowance set on the owner's account for the spender.
+ *
+ * @param operator - address of the operator
+ * @param tokenId - tokenId of the token
+ * @returns
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:150]: retursn true if ...

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 15]
```diff
@@ -1,282 +1,239 @@
...
-  assert(!_isApproved(spenderAddress, tokenId), 'Already approved');
-
-  const key = approvedTokenKey + tokenId.toString();
-  Storage.set(key, spenderAddress);
+export function _balanceOf(owner: string): u256 {
+  const key = balanceKey(owner);
+  if (Storage.has(key)) {
+    return bytesToU256(Storage.get(key));
+  }
+  return u256.Zero;
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:79]: Any reason for not using ternary operator for such simple instructions?
```suggestion
return Storage.has(key) ? bytesToU256(Storage.get(key)) : u256.Zero;
```

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 16]
```diff
@@ -1,282 +1,239 @@
...
+  return (
+    _isApproved(operator, tokenId) ||
+    _isApprovedForAll(_ownerOf(tokenId), operator) ||
+    _ownerOf(tokenId) == operator
   );
 }
 
-export function assertAddressIsValid(address: string): void {
-  assert(validateAddress(address), 'Address is not valid');
+export function _update(to: string, tokenId: u256, auth: string): void {
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:204]: No documentation

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 17]
```diff
@@ -1,282 +1,239 @@
...
+  const from = _ownerOf(tokenId);
+  if (auth != '') {
+    assert(_isAuthorized(auth, tokenId), 'Unauthorized');
+  }
+  if (from != '') {
+    // clear the approval
+    _approve('', tokenId);
+    // update the balance of the from
+    const fromBalance = bytesToU256(Storage.get(balanceKey(from)));
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:213]: You're not checking that the balance is not 0

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 18]
```diff
@@ -1,282 +1,239 @@
...
+    Storage.set(balanceKey(from), u256ToBytes(fromBalance - u256.One));
+  }
+  if (to != '') {
+    const toBalanceKey = balanceKey(to);
+    // update the balance of the to
+    if (Storage.has(toBalanceKey)) {
+      Storage.set(
+        toBalanceKey,
+        u256ToBytes(bytesToU256(Storage.get(toBalanceKey)) + u256.One),
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:222]: You're not checking that the balance is not u256.Max

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 19]
```diff
@@ -1,282 +1,239 @@
...
 }
 
-export function assertAddressIsValid(address: string): void {
-  assert(validateAddress(address), 'Address is not valid');
+export function _update(to: string, tokenId: u256, auth: string): void {
+  const from = _ownerOf(tokenId);
+  if (auth != '') {
+    assert(_isAuthorized(auth, tokenId), 'Unauthorized');
+  }
+  if (from != '') {
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:209]: Can the 'from' field be empty? In that case, does it mean that the 'to' field would receive a token out of nowhere?

File: smart-contracts/assembly/contracts/NFT/__tests__/NFT-internals.spec.ts [hunk 20]
```diff
@@ -0,0 +1,96 @@
+import { resetStorage, setDeployContext } from '@massalabs/massa-as-sdk';
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/__tests__/NFT-internals.spec.ts:1]: Can you add edge cases unit tests?

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts [hunk 21]
```diff
@@ -1,282 +1,239 @@
...
-}
-
-export function _getBalanceOf(address: string): u64 {
-  const balanceKey = stringToBytes('balanceOf_' + address);
-
-  if (!Storage.has(balanceKey)) return 0;
-
-  return bytesToU64(Storage.get(balanceKey));
+export function _constructor(name: string, symbol: string): void {
+  Storage.set(NAME_KEY, stringToBytes(name));
```
Comment (gregLibert) [inline smart-contracts/assembly/contracts/NFT/NFT-internals.ts:26]: Why did you removed the `assert(Context.isDeployingContract());`? I have the feeling that now we can change the symbol and rename the the nft as we want.

**Review by gregLibert — CHANGES_REQUESTED** [review]
Not sure I'm the best person to review that kind of thing. 

@leoloco , let's sync on this before modifying anything; I don't want you to waste your time.

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts (hunk 10 above)
Comment (leoloco) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:123]: Can you give an example ? 
I guess you want me to forward the error to the top level call but I don't know how its done

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts (hunk 12 above)
Comment (leoloco) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:128]: yes, validation is to be done by the wrapper

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts (hunk 17 above)
Comment (leoloco) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:213]: Balance cannot be zero because from is `_ownerOf(tokenId)`

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts (hunk 19 above)
Comment (leoloco) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:209]: Yes indeed. I've added explainations

File: smart-contracts/assembly/contracts/NFT/NFT-internals.ts (hunk 21 above)
Comment (leoloco) [reply smart-contracts/assembly/contracts/NFT/NFT-internals.ts:26]: My idea is that this is not the final constructor but rather an helper that should be wrapped into the actual constructor. Thus, the `assert(Context.isDeployingContract());` isn't mandatory as it should be included in the final constructor

**Review by gregLibert — APPROVED** [review]
